extra/src/main/java/aima/extra/environment/model/map/object/MapObject.java

```diff
@@ -0,0 +1,82 @@
+package aima.extra.environment.model.map.object;
+
+/**
+ * Created by @AdrianBZG on 2/04/17.
+ */
+
+import aima.core.util.datastructure.Tuple;
+import aima.extra.environment.view.object.ObjectView;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+
+public abstract class MapObject implements Cloneable {
```

Comment:
I'd go with a component based approach, but that's just an opinion. http://gameprogrammingpatterns.com/component.html
this would remove mostly anything from the `MapObject` class and instead place some logic in the interfaces and create a `class Obstacle extends MapObject inplements Position, Render, ReadOnlyOptions ...`.